## Упражнения по библиотеке Numpy

In [1]:
import numpy as np
import random

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

**1.** Given a random array, change the sign of elements with values between 3 and 8.

In [2]:
rand_arr = np.random.randint(0,10, size=10)
print("Original Array:", rand_arr)

rand_arr[(rand_arr >=3) & (rand_arr <=8)] = -rand_arr[(rand_arr >=3) & (rand_arr <=8)]
print("Modified Array:", rand_arr)

Original Array: [7 5 1 4 0 5 1 9 5 5]
Modified Array: [-7 -5  1 -4  0 -5  1  9 -5 -5]


**2.** Заменить максимальный элемент случайного массива на 0

**2.** Replace the maximum element of a random array with 0.

In [3]:
rand_arr = np.random.randint(0,10, size=10)
print("Original Array:", rand_arr)

rand_arr[rand_arr == np.max(rand_arr)] = 0
print("Modified Array:", rand_arr)



Original Array: [2 2 1 2 9 3 8 3 3 6]
Modified Array: [2 2 1 2 0 3 8 3 3 6]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

**3.** Construct the Cartesian product of arrays (all combinations of every element). The input is a two-dimensional array.

In [4]:
def cartesian_product(arr):
  grid = np.meshgrid(*arr,indexing='ij')
  return np.stack(grid, axis=-1).reshape(-1, len(arr))

rand_arr = np.random.randint(0,10, size=(2,4))
print("Original Array:", rand_arr)

print("Cartesian Product:", cartesian_product(rand_arr))


Original Array: [[2 3 0 0]
 [3 1 7 6]]
Cartesian Product: [[2 3]
 [2 1]
 [2 7]
 [2 6]
 [3 3]
 [3 1]
 [3 7]
 [3 6]
 [0 3]
 [0 1]
 [0 7]
 [0 6]
 [0 3]
 [0 1]
 [0 7]
 [0 6]]


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

**4.** Given two arrays A (8x3) and B (2x2). Find the rows in A that contain elements from every row in B, regardless of the order of elements in B

In [5]:
A = np.random.randint(0,10,size=(8,3))
B = np.random.randint(0,10,size=(2,2))

print("Original Array A:", A)
print("Original Array B:", B)
B0 = set(B[0])
B1 = set(B[1])

mask1 = np.isin(A,list(B0)).any(axis=1)
mask2 = np.isin(A,list(B1)).any(axis=1)

combined_mask = np.logical_and(mask1,mask2)

print("Selected Rows Of A:", A[combined_mask])


Original Array A: [[9 2 1]
 [0 7 5]
 [1 5 2]
 [5 8 5]
 [2 2 9]
 [8 2 6]
 [4 7 3]
 [8 4 7]]
Original Array B: [[9 9]
 [4 1]]
Selected Rows Of A: [[9 2 1]]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

**5.** Given a 10x3 matrix, find the rows consisting of non-equal values (e.g., row [2,2,3] remains, row [3,3,3] is removed)

In [6]:

A = np.random.randint(0, 2, size=(10, 3))
print("Original Array:", A)

first_column = A[:, 0:1]
is_equal_to_first = (A == first_column)
all_same = is_equal_to_first.all(axis=1)

mask = ~all_same
result = A[mask]
print("Result:",result)

Original Array: [[0 0 1]
 [0 0 1]
 [1 1 0]
 [1 0 0]
 [0 1 1]
 [1 0 0]
 [0 1 0]
 [1 0 0]
 [0 0 1]
 [1 1 1]]
Result: [[0 0 1]
 [0 0 1]
 [1 1 0]
 [1 0 0]
 [0 1 1]
 [1 0 0]
 [0 1 0]
 [1 0 0]
 [0 0 1]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

**6.** Given a two-dimensional array. Remove rows that are duplicates.

In [7]:
arr = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [1, 2, 3], # Duplicate of row 0
    [7, 8, 9],
    [4, 5, 6]  # Duplicate of row 1
])
print("Original Array:", arr)

unique_arr = np.unique(arr,axis=0)
print("Modified Array:", unique_arr)

Original Array: [[1 2 3]
 [4 5 6]
 [1 2 3]
 [7 8 9]
 [4 5 6]]
Modified Array: [[1 2 3]
 [4 5 6]
 [7 8 9]]


______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).

Dual-Implementation Tasks (1-5)
For each of the following tasks, provide two implementations: one without using NumPy (assume inputs/outputs are standard Python lists), and a second fully vectorized implementation using NumPy (no Python loops/map/list comprehensions).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

**Note 1.** Assume all objects are non-empty (e.g., in Task 1, there are non-zero elements on the diagonal).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

**Note 2.** For most tasks, the solution takes no more than 1-2 lines.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

* __Task 1__: Calculate the product of non-zero elements on the diagonal of a
rectangular matrix. Example: For X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]), the answer is 3.

In [11]:
#Python Implementation
Product = 1
X = np.array([
    [1,0,1],
    [2,0,2],
    [3,0,3],
    [4,4,4]
])
for indexI, valI in enumerate(X):
      for indexJ, valJ in enumerate(valI):
          if indexI == indexJ:
              if X[indexI][indexJ] != 0:
                  Product *= X[indexI][indexJ]
print(Product)



3


In [10]:
#Numpy Implementation
X = np.array([
    [1,0,1],
    [2,0,2],
    [3,0,3],
    [4,4,4]
])
Diagonal = np.diag(X)
Product = np.prod(Diagonal[Diagonal != 0])
print(Product)

3


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

* __Task 2__: Given two vectors x and y. Check if they define the same multiset. Example: For x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]), the answer is True.

In [16]:
#Python Implementation
x = np.array([1,2,2,4])
y = np.array([4,2,1,2])

if len(x) != len(y):
  print("False")
else:
  print(sorted(x) == sorted(y))


True


In [18]:
#Numpy Implementation
x = np.array([1,2,2,4])
y = np.array([4,2,1,2])

if x.shape != y.shape:
  print("False")
else:
  print(np.array_equal(np.sort(x),np.sort(y)))



True


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль.
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

 * __Task 3__: Find the maximum element in vector x among elements preceded by a zero. Example: For x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]), the answer is 5.

In [29]:
#Python Implementation
x = np.array([6,2,0,3,0,0,5,7,0])

max_element = 0
for index, val in enumerate(x[1:],start=1):
  if x[index-1] == 0:
    if val > max_element:
      max_element = val
    else:
      continue
print(max_element)

5


In [31]:
#Numpy Implementation
x = np.array([6,2,0,3,0,0,5,7,0])
mask = (x[:-1] == 0)
elements_after_zero = x[1:][mask]
max_element = elements_after_zero.max() if elements_after_zero.size > 0 else 0

print(max_element)

5


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

 * __Task 4__: Implement Run-length encoding (RLE). For a vector x, return a tuple of two arrays of equal length: the first contains the values, and the second contains their counts. Example: For x = np.array([2, 2, 2, 3, 3, 3, 5]), the answer is (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [36]:
#Python Implementation
def rle_python(x):
    if len(x) == 0:
        return np.array([]), np.array([])

    values = []
    counts = []

    current_val = x[0]
    count = 0

    for val in x:
        if val == current_val:
            count += 1
        else:
            values.append(current_val)
            counts.append(count)
            current_val = val
            count = 1

    values.append(current_val)
    counts.append(count)

    return np.array(values), np.array(counts)

vals, counts = rle_python(x)
print(vals, counts)


[2 3 5] [3 3 1]


In [39]:
#Numpy Implementation
x = np.array([2,2,2,3,3,3,5])
def rle_numpy(x):
    if len(x) == 0:
        return np.array([]), np.array([])

    locs = np.where(x[1:] != x[:-1])[0]
    boundaries = np.concatenate(([ -1], locs, [len(x) - 1]))

    counts = np.diff(boundaries)
    values = x[boundaries[1:]]

    return values, counts

vals, counts = rle_numpy(x)
print(vals, counts)

[2 3 5] [3 3 1]


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

* __Task 5__: Given two samples of objects - X and Y. Calculate the matrix of Euclidean distances between objects. Compare the performance with the scipy.spatial.distance.cdist function.

In [42]:
#Python Implementation
import math
import time
from scipy.spatial.distance import cdist

def python_euclidean_dist(X, Y):
    num_x = len(X)
    num_y = len(Y)
    dist_matrix = [[0.0 for _ in range(num_y)] for _ in range(num_x)]

    for i in range(num_x):
        for j in range(num_y)
            squared_sum = 0.0
            for k in range(len(X[i])):
                squared_sum += (X[i][k] - Y[j][k])**2
            dist_matrix[i][j] = math.sqrt(squared_sum)
    return dist_matrix

#Numpy Implementation
def numpy_euclidean_dist(X, Y):
    x_2 = np.sum(X**2, axis=1, keepdims=True)
    y_2 = np.sum(Y**2, axis=1)

    xy = 2 * np.dot(X, Y.T)
    return np.sqrt(np.maximum(x_2 - xy + y_2, 0))


# Testing
N, M, D = 300, 300, 50
X_np = np.random.rand(N, D)
Y_np = np.random.rand(M, D)

# Convert to lists for the Python implementation test to be accuratee
X_list = X_np.tolist()
Y_list = Y_np.tolist()

print(f"Testing distance matrix calculation ({N}x{M})...\n")

# 1. Pure Python
start = time.time()
res_py = python_euclidean_dist(X_list, Y_list)
t_py = time.time() - start
print(f"Python Implementation: {t_py:.5f} seconds")

# 2. NumPy
start = time.time()
res_np = numpy_euclidean_dist(X_np, Y_np)
t_np = time.time() - start
print(f"NumPy Implementation:  {t_np:.5f} seconds")

# 3. SciPy cdist
start = time.time()
res_sp = cdist(X_np, Y_np, 'euclidean')
t_sp = time.time() - start
print(f"SciPy cdist:           {t_sp:.5f} seconds")


Testing distance matrix calculation (300x300)...

Python Implementation: 0.63604 seconds
NumPy Implementation:  0.00273 seconds
SciPy cdist:           0.00449 seconds


_______
________

* #### __Задача 6__: CrunchieMunchies __*__
* #### __Task 6__: CrunchieMunchies__ *__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

You work in the marketing department of the food company MyCrunch, which is developing a new type of tasty, healthy cereal called **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

You want to demonstrate to consumers how healthy your cereal is compared to other leading brands, so you have collected nutritional data for several different competitors.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.

Your task is to use NumPy calculations to analyze this data and prove that your CrunchieMunchies are the healthiest choice.


In [ ]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

1. Review the file cereal.csv. This file contains calorie counts for various cereal brands. Load the data from the file and save it as calorie_stats.

In [49]:
calorie_stats = np.loadtxt("/content/pandas_numpy_tasks/data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Average Comparison: One serving of CrunchieMunchies contains 60 calories.

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

How much higher is the average calorie count of your competitors? Save the answer in average_calories and print it.

In [55]:
crunchie_munchies_calories = 60
average_calories = np.mean(calorie_stats)
diff = average_calories - crunchie_munchies_calories
average_calories, diff

(np.float64(106.88311688311688), np.float64(46.883116883116884))

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Distribution Check: Does the average calorie count accurately reflect the dataset's distribution?

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

Let's sort the data and see. Sort the data and save the result in calorie_stats_sorted. Print the sorted info.

In [62]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Median: It seems most values are above the average. Let's see if the median is a more accurate indicator.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

Calculate the median and save it in median_calories. Print it to compare with the average.

In [63]:
median_calories = np.median(calorie_stats_sorted)
median_calories, average_calories

(np.float64(110.0), np.float64(106.88311688311688))

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Percentiles: While the median shows at least half the values are over 100 calories, it would be more impressive to show that a significant portion of competitors has higher calories than CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

Calculate different percentiles until you find the lowest percentile that exceeds 60 calories. Save this in nth_percentile.

In [66]:
nth_pernctile = None
for percentile in range(0,101):
  current = np.percentile(calorie_stats_sorted, percentile)
  print(current)
  if current > 60:
    nth_pernctile = current
    break
nth_pernctile


50.0
50.0
50.0
55.599999999999994
70.0


np.float64(70.0)

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Percentage Comparison: While percentiles are useful, they can be awkward for marketing.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

Instead, calculate the percentage of cereals that contain more than 60 calories per serving. Save this in more_calories and print it.

In [74]:
calorie_stats_sorted_above_60 = calorie_stats_sorted[calorie_stats_sorted > 60]
calorie_stats_sorted_above_60
more_calories = (calorie_stats_sorted_above_60.size / calorie_stats_sorted.size) * 100
more_calories

96.1038961038961

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

How much variation is in the dataset? Can we generalize that most cereals have around 100 calories, or is the spread larger?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

Calculate the standard deviation and save it in calorie_std.

In [75]:
calorie_std = np.std(calorie_stats_sorted)
calorie_std

np.float64(19.35718533390827)

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Write a short paragraph summarizing your findings and how you believe this data can be used to MyCrunch's advantage when marketing CrunchieMunchies.

Based on the comparative analysis, CrunchieMunchies hold a significant nutritional advantage in the marketplace, with only **60** calories per serving, it sits below the competitor average of **106** calories and median of **110** calories. Most Importantly **96%** of competing products have a higher calorie count than CrunchieMunchies.